scripts: kconfig: Turn most warnings into errors

ulfalizer opened the pull request:
In particular, this will turn assignments to undefined Kconfig variables
into errors, which are very easy to miss otherwise (e.g. when Kconfig
symbols get renamed or removed).

Warnings generated by anything tested by CI (scripts/sanitycheck) will
get caught.

Have a whitelist of warnings that are not turned into errors. Some
warnings currently whitelisted should be turned into errors as well, but
would require a bit of work.

ulfalizer committed:
scripts: kconfig: Turn most warnings into errors

In particular, this will turn assignments to undefined Kconfig variables
into errors, which are very easy to miss otherwise (e.g. when Kconfig
symbols get renamed or removed).

Warnings generated by anything tested by CI (scripts/sanitycheck) will
be caught.

Have a whitelist of warnings that are not turned into errors. Some
warnings currently whitelisted should be turned into errors as well, but
would require a bit of work.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer committed:
tests: Do not build arm_irq_vector_table .config's for ARC

Previously, ARC .config's would get built for this test, giving Kconfig
warnings along the lines of

  warning: NUM_IRQS ... was assigned the value "3" but got the value
  "38". Check its dependencies...

Add an 'arch_whitelist: arm' so that .configs only get built for ARM.

These warnings will soon be turned into errors.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer commented:
CC @carlescufi @SebastianBoe @nashif

SebastianBoe commented on scripts/kconfig/kconfig.py:
```diff
@@ -5,6 +5,35 @@
+
+    # Warning generated when a symbol is set more than once in a .config file.
+    # These should be investigated, but whitelist them for now.
+    "set more than once",
```
Merging two different Kconfig fragments with overlapping and agreeable requirements should not produce a warning.

ulfalizer replied:
If someone messes up and puts in the same assignment twice, it's cool to point it out.

There are some cases where it makes sense to have more than one assignment to the same variable though, when merging different combinations of `.config` files. Might even be okay to override assignments from the base configuration with a different value in some cases.

It's a tradeoff. If you think those redundant assignment warnings (when you give the same value both times) are more annoying than helpful, I could disable them though.

SebastianBoe replied:
I do. Please disable.

ulfalizer committed:
samples: bluetooth: Fix microbit/nrf5 UART flow control assignments

The UART_NRFX_FLOW_CONTROL Kconfig symbol was renamed to
UART_0_NRF_FLOW_CONTROL in commit 3f99eefe5af74 ("drivers: uart: Rename
nrf5 namings to nrfx").

The assignments in samples/bluetooth/hci_uart/microbit.conf and
samples/bluetooth/hci_uart/nrf5.conf were updated in the same commit,
but to the wrong name. Change them to use the correct name.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer commented:
CC @jarz-nordic (3f99eef)

ulfalizer commented:
@inakypg @nashif @andrewboie 
What's the purpose of the `CONFIG_NUM_IRQS=2` assignment in `lib/sprintf/prj.conf`? Is it needed?

It's causing a warning for ARC, where `NUM_IRQS` [has a `range`]( http://docs.zephyrproject.org/reference/kconfig/CONFIG_NUM_IRQS.html#kconfig-definitions):

    Using /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/boards/arc/em_starterkit/em_starterkit_em7d_defconfig as base
    Merging /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/tests/lib/sprintf/prj.conf
    ...
    Error: Aborting due to non-whitelisted Kconfig warning 'warning: user value 2 on the int symbol NUM_IRQS (defined at ...) ignored due to being outside the active range ([17, 256]) -- falling back on defaults'.

From https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/17015/1/console.

andrewboie commented:
> What's the purpose of the CONFIG_NUM_IRQS=2 assignment in lib/sprintf/prj.conf? Is it needed?

Cruft. I sent #8347 .

ulfalizer commented:
recheck
